### ISSUE TITLE
Performance issue on chunk generation

### ISSUE THREAD

**Issue by Gwani** [keepcalm/BukkitForge#532]
Background: 
Our test dedicated server had some severe lag issues when generating new chunks (eg. when teleporting to unexplored coordinates) which caused clients to get disconnected due to a timeout. We're using a lot of mods which hook into chunk generation (ores etc.), so i investigated the problem by looking at the CPU time spent executing various Java methods using Visual VM. We could fix the disconnect problems by removing a less important mod which managed to get itself to the top of the list during chunk generation. However, second on the list was the function keepcalm.mods.events.ForgeEventHelper.onBlockBreak(), so we investigated further.

Issue:
During chunk generation on a dedicated server, the Java VM spends quite a lot of CPU time calling keepcalm.mods.events.ForgeEventHelper.onBlockBreak().

Reproducing the issue:
We tested the issue on a clean server installation running minecraftforge-universal-1.4.7-6.6.0.497 and BukkitForge-1.4.7-244. No additional mods and no Bukkit plugins were installed. Performance analaysis was done using VisualVM. We logged onto the server, teleported to 100000 100 100000 and sampled CPU time for the server process during chunk generation. The top 3 functions sampled using VisualVM were:

keepcalm.mods.events.ForgeEventHelper.onBlockBreak() - 42%
in.a() - 12%
kc.a() - 4%

Conclusion:
Block break events are expected to happen right after chunk generation (sand/gravel settling, water flowing onto lava, etc.), however, the amount of CPU time for onBlockBreak() suggests that the function is either called more often than necessary (maybe the function is called for every newly generated block? also: no plugins installed, so no plugins need to receive the block break event), or it introduces quite some overhead. 
The issue is not serious, but it might be worth investigating - on a server with a lot of players that CPU time is better spent on actual world generation.

**Comment by Dreadcor** [comment]
Hi , i was wondering if you can show me how you did this step by step if possible?

**Comment by Gwani** [comment]
Hi there! I assume you refer to the profiling with VisualVM. You can get VisualVM at http://visualvm.java.net, its a quite useful tool when it comes to examining the runtime behaviour of a Java application. I used its sampling profiler to find the method it question. If you run a Java apllication such as Minecraft, it appears in the Applications list on the left under "Local". Right click and select "Sample". This opens the sampler tab which has 3 buttons: "CPU", "Memory" and "Stop". We're interested in CPU sampling, so we start it by pressing "CPU". VisualVM now queries the Java VM at regular intervals for the time it spends executing the various functions of the Java apllication and compiles a histogram table. You can easily see how much time is spent in each function. Note that "Self time" is the total time spent in a function including the a thread is idle and not using any CPU time. So we're really interested in "Self time (CPU)", thats the time the function actually performs work.

You should also note that the sampling profiler is only useful for finding the hot spots in an application, those are the functions which have greatest performance impact. VisualVM also has an instrumenting profiler ("Profiler" tab) which allows for in-depth analysis of the function calls, to find out where inside the function the time is actially spent. You should read the VisualVM documentation and find some tutorials on how to use it if you're interested in its other features.

**Comment by Dreadcor** [comment]
Thanks, ill see if i can get some of this info with the issues i have :)

**Comment by gravypod** [comment]
@SpoonsJTD its the reflection @keepcalm's been doing, we need to replace it with your monkey patching

**Comment by SpoonsJTD** [comment]
Would like to see this rerun. Was switched to use throwable().getStackTrace and got rid of threading and sleeping. Need to reimplement using sun internal class to get even faster.

**Comment by Gwani** [comment]
@SpoonsJTD I just tested it again with BukkitForge-1.4.7-255: keepcalm.mods.events.ForgeEventHelper.onBlockBreak() is still the top function call with almost 50% self time (CPU), but chunk generation runs a lot faster now. On my local test server its even fast enough to generate chunks in time when i'm flying in creative mode (before, i had to wait about 20 seconds at the chunk boundary to see the newly generated chunks).

However, you might still want to schedule block break handling for a reimplementation -  it should have an almost nonexistent performance impact, especially when no plugins are loaded to receive the event.

**Comment by Ionaru** [comment]
I am having some problems with this on my server. When people are exploring and generating new chunks there is a great drop in TPS. This is very annoying for others playing on the server.

**Comment by WhiplashD** [comment]
Yup I've had performance issues on my server as well when either I or someone else generates new chunks. Other than the occasional can't keep up error message though I haven't experienced anything bad from it.

**Comment by SpoonsJTD** [comment]
I made another pass on the block break implementation. It can't be negligible unfortunately because we have to do some checking into who called block break to understand whether to fire the event or not.

**Comment by Gwani** [comment]
@SpoonsJTD Forgive me if i'm making stupid suggestions without knowing the whole code, i'm just trying to be helpful: I wonder if you have some sort of control over the code which calls the block break leading to an event to be fired (or alternatively the code making it NOT fire)? Perhaps its possible to augment the call you receive with a flag which tells you whether to fire or not (?). You would then just have to check a boolean or a single bit, instead of doing multiple string comparisons.

I just took a peek at the revised function and see that you got rid of the loop iterating all stack trace elements and eliminated the redundant string conversions to lower case. Well done, that should run quite a bit faster now. However, in line 277 you still generate a stack trace which doesn't seem to be used anywhere. You might also want to get rid of that, i believe getStackTrace() is rather costly in comparison to the other functions used.

 I'll do another test later this weekend, Thanks for your efforts!
Gwani

**Comment by SpoonsJTD** [comment]
Well that was dumb. Went to all the trouble to make the call perform better and left in the call that eats all the resources. 

It's not a stupid suggestion, btw, and if we had complete control over all the code that goes into making a block break call, we would do it much like you describe. Unfortunately, we don't, so we have to rely on checking to see who is calling the function and how they are calling it to check if there should be an event. 

Theoretically, any system where the Sun reflection method is available should run much, much faster. Nallar provided a link to the implementation and a profiled test of it, and the new call I am using executes at an order of magnitude or more faster than any of the previous methods I've used. Build 257 will have the resource hog line removed.

**Comment by Gwani** [comment]
Just tested build 257 on local server. Great work! Performance impact of keepcalm.mods.events.ForgeEventHelper.onBlockBreak() is now not measurable for my test scenario: 0% self time (CPU). Thanks again!
